src/core/sim.py

```diff
@@ -30,8 +30,13 @@ def step(self) -> PropagatedOutput:
         for actuator_model in self._models.actuator:
             self.state.state.update(actuator_model.evaluate(self.state.state))
 
+        # Propagate derived state
+        for derived_state_model in self._models.derived:
+            self.state.state.derived_state.update(
+                derived_state_model.evaluate(self.state.time, self.state.state))
+
```

Comment:
I think we might have to run this at every initialization of a `State` class instead of in the main loop. Because a derived state is directly linked to a state

Reply:
I believe doing so would result in State depending on Model and Model likewise depending on State— not sure how we could go about it without circular references/imports

Reply:
Good point. The way i'd go around this is to make derived models completely independent of the `Model` class - maybe we could keep this strictly functional and just cut/paste the functions from Derived Models to state.py